File: Python/Binary Tree BFS/Cousins in Binary Tree II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        levelsum=[]
        q=[]
        q.append(root)
        #1st: bfs
        while len(q) !=0:
            size=len(q)
            sum=0
            for i in range(size):
                ele=q.pop(0)
                sum+=ele.val
                if ele.left:
                    q.append(ele.left)
                if ele.right:
                    q.append(ele.right)
            levelsum.append(sum)
        #2nd:bfs
        q2=[]
        q2.append(root)
        level =1
        root.val=0
        while len(q2) !=0:
            s=len(q2)
            for i in range(s):
                curr=q2.pop(0)
                siblingsum = 0
                if curr.left:
                    siblingsum+=curr.left.val
                if curr.right:
                    siblingsum+=curr.right.val
                if curr.left:
                    curr.left.val=levelsum[level]-siblingsum
                    q2.append(curr.left)
                if curr.right:
                    curr.right.val=levelsum[level]-siblingsum
                    q2.append(curr.right)
            level+=1
        return root
        

File: CPP/Binary Tree BFS/Cousins in Binary Tree II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* replaceValueInTree(TreeNode* root) {
        queue<TreeNode*>bfsq;
        bfsq.push(root);
        vector<int>levelsum;
        while(!bfsq.empty()){
            int sum = 0;
            int size = bfsq.size();
            for(int i=0;i<size;i++){
                TreeNode* node= bfsq.front();
                bfsq.pop();
                sum+=node->val;
                if(node->left)bfsq.push(node->left);
                if(node->right)bfsq.push(node->right);
            }
            levelsum.push_back(sum);
        }
        queue<TreeNode*>q;
        q.push(root);
        int level = 1;
        root->val=0;
        while(!q.empty()){
            int s= q.size();
            for(int i=0;i<s;i++){
                TreeNode* curr = q.front();
                q.pop();
                int siblingsum = 0;
                if(curr->left){
                    siblingsum+=curr->left->val;
                }
                if(curr->right){
                    siblingsum+=curr->right->val;
                }
                if(curr->left){
                    curr->left->val=levelsum[level]-siblingsum;
                    q.push(curr->left);
                }
                if(curr->right){
                    curr->right->val=levelsum[level]-siblingsum;
                    q.push(curr->right);
                }
            }
            level++;
        }
        return root;
    }
};
